File: public/cpp/069-permutations-ii.cpp
/*
 * Problem: Permutations II (LeetCode 47)
 * Difficulty: Medium
 * Category: Backtracking and Recursion
 * 
 * Description:
 * Given a collection of numbers, nums, that might contain duplicates, return all
 * possible unique permutations in any order.
 * 
 * Example 1:
 * Input: nums = [1,1,2]
 * Output: [[1,1,2],[1,2,1],[2,1,1]]
 * 
 * Example 2:
 * Input: nums = [1,2,3]
 * Output: [[1,2,3],[1,3,2],[2,1,3],[2,3,1],[3,1,2],[3,2,1]]
 * 
 * Constraints:
 * - 1 <= nums.length <= 8
 * - -10 <= nums[i] <= 10
 */

#include <iostream>
#include <vector>
#include <algorithm>
#include <unordered_set>
#include <set>
using namespace std;

/*
 * APPROACH 1: BACKTRACKING WITH SORTING AND SKIP DUPLICATES
 * 
 * Intuition:
 * - Sort array to group duplicates together
 * - Use backtracking with a used array to track which elements are used
 * - Skip duplicate elements at the same recursion level
 * - Key insight: if nums[i] == nums[i-1] and nums[i-1] is not used,
 *   skip nums[i] to avoid duplicate permutations
 * 
 * Algorithm:
 * 1. Sort the array
 * 2. Use backtracking with used array
 * 3. For each position, try each unused element
 * 4. Skip if current element equals previous and previous is not used
 * 5. This ensures we only use duplicates in order
 * 
 * Time Complexity: O(n! * n) - n! permutations, O(n) to copy each
 * Space Complexity: O(n) - recursion depth + used array
 */
class Solution1 {
private:
    void backtrack(vector<int>& nums, vector<bool>& used, 
                   vector<int>& current, vector<vector<int>>& result) {
        // Base case: permutation complete
        if (current.size() == nums.size()) {
            result.push_back(current);
            return;
        }
        
        // Try each element
        for (int i = 0; i < nums.size(); i++) {
            // Skip if already used
            if (used[i]) continue;
            
            // Skip duplicates: if current equals previous and previous not used
            // This ensures we use duplicates in order
            if (i > 0 && nums[i] == nums[i-1] && !used[i-1]) {
                continue;
            }
            
            // Use nums[i]
            used[i] = true;
            current.push_back(nums[i]);
            
            backtrack(nums, used, current, result);
            
            // Backtrack
            current.pop_back();
            used[i] = false;
        }
    }
    
public:
    vector<vector<int>> permuteUnique(vector<int>& nums) {
        vector<vector<int>> result;
        vector<int> current;
        vector<bool> used(nums.size(), false);
        
        // Sort to group duplicates
        sort(nums.begin(), nums.end());
        
        backtrack(nums, used, current, result);
        return result;
    }
};

/*
 * APPROACH 2: BACKTRACKING WITH SWAP AND SET
 * 
 * Intuition:
 * - Use swap-based backtracking
 * - At each position, use a set to track which values we've already tried
 * - This prevents generating duplicate permutations
 * 
 * Algorithm:
 * 1. For each position, maintain a set of used values
 * 2. Try swapping with each position from current to end
 * 3. Skip if value already tried at this position
 * 4. Recursively permute remaining positions
 * 
 * Time Complexity: O(n! * n)
 * Space Complexity: O(n^2) - recursion depth * set at each level
 */
class Solution2 {
private:
    void backtrack(vector<int>& nums, int start, vector<vector<int>>& result) {
        if (start == nums.size()) {
            result.push_back(nums);
            return;
        }
        
        // Track which values we've tried at this position
        unordered_set<int> used;
        
        for (int i = start; i < nums.size(); i++) {
            // Skip if we've already tried this value at this position
            if (used.count(nums[i])) continue;
            
            used.insert(nums[i]);
            
            // Swap and recurse
            swap(nums[start], nums[i]);
            backtrack(nums, start + 1, result);
            swap(nums[start], nums[i]); // Backtrack
        }
    }
    
public:
    vector<vector<int>> permuteUnique(vector<int>& nums) {
        vector<vector<int>> result;
        backtrack(nums, 0, result);
        return result;
    }
};

/*
 * APPROACH 3: FREQUENCY MAP BACKTRACKING
 * 
 * Intuition:
 * - Count frequency of each unique element
 * - Build permutations by choosing from available elements
 * - Decrement frequency when using an element
 * - This naturally handles duplicates
 * 
 * Algorithm:
 * 1. Build frequency map
 * 2. For each position, try each unique element with available count
 * 3. Decrement count, recurse, then restore count
 * 4. No need to sort or track used array
 * 
 * Time Complexity: O(n! * n)
 * Space Complexity: O(n) - frequency map + recursion
 */
class Solution3 {
private:
    void backtrack(unordered_map<int, int>& freq, int n,
                   vector<int>& current, vector<vector<int>>& result) {
        if (current.size() == n) {
            result.push_back(current);
            return;
        }
        
        // Try each unique element
        for (auto& [num, count] : freq) {
            if (count > 0) {
                // Use this element
                current.push_back(num);
                freq[num]--;
                
                backtrack(freq, n, current, result);
                
                // Backtrack
                current.pop_back();
                freq[num]++;
            }
        }
    }
    
public:
    vector<vector<int>> permuteUnique(vector<int>& nums) {
        // Build frequency map
        unordered_map<int, int> freq;
        for (int num : nums) {
            freq[num]++;
        }
        
        vector<vector<int>> result;
        vector<int> current;
        backtrack(freq, nums.size(), current, result);
        
        return result;
    }
};

/*
 * APPROACH 4: ITERATIVE WITH INSERTION
 * 
 * Intuition:
 * - Build permutations iteratively by inserting each element
 * - For each new element, insert it at all possible positions in existing permutations
 * - Use set to avoid duplicates
 * 
 * Algorithm:
 * 1. Start with empty permutation
 * 2. For each element, insert it at all positions in all existing permutations
 * 3. Use set to automatically handle duplicates
 * 4. Convert set to vector at the end
 * 
 * Time Complexity: O(n! * n^2) - insertion takes O(n)
 * Space Complexity: O(n! * n) - set storage
 */
class Solution4 {
public:
    vector<vector<int>> permuteUnique(vector<int>& nums) {
        set<vector<int>> uniquePerms;
        vector<vector<int>> current = {{}};
        
        for (int num : nums) {
            vector<vector<int>> next;
            
            for (const auto& perm : current) {
                // Insert num at all positions
                for (int i = 0; i <= perm.size(); i++) {
                    vector<int> newPerm = perm;
                    newPerm.insert(newPerm.begin() + i, num);
                    next.push_back(newPerm);
                }
            }
            
            current = next;
        }
        
        // Remove duplicates using set
        for (const auto& perm : current) {
            uniquePerms.insert(perm);
        }
        
        return vector<vector<int>>(uniquePerms.begin(), uniquePerms.end());
    }
};

/*
 * APPROACH 5: NEXT PERMUTATION APPROACH
 * 
 * Intuition:
 * - Sort array to get first permutation
 * - Use next_permutation algorithm repeatedly
 * - This generates permutations in lexicographic order
 * - Automatically handles duplicates
 * 
 * Algorithm:
 * 1. Sort array
 * 2. Add current permutation to result
 * 3. Generate next permutation using STL or custom implementation
 * 4. Repeat until no more permutations
 * 
 * Time Complexity: O(n! * n) - n! permutations, O(n) for next_permutation
 * Space Complexity: O(1) - excluding output
 */
class Solution5 {
public:
    vector<vector<int>> permuteUnique(vector<int>& nums) {
        vector<vector<int>> result;
        
        // Sort to get first permutation
        sort(nums.begin(), nums.end());
        
        do {
            result.push_back(nums);
        } while (next_permutation(nums.begin(), nums.end()));
        
        return result;
    }
};

// Helper function to print permutations
void printPermutations(const vector<vector<int>>& perms) {
    cout << "[";
    for (int i = 0; i < perms.size(); i++) {
        cout << "[";
        for (int j = 0; j < perms[i].size(); j++) {
            cout << perms[i][j];
            if (j < perms[i].size() - 1) cout << ",";
        }
        cout << "]";
        if (i < perms.size() - 1) cout << ",";
    }
    cout << "]\n";
}

// Test function
void test(vector<int> nums, int approach) {
    vector<vector<int>> result;
    
    cout << "Input: [";
    for (int i = 0; i < nums.size(); i++) {
        cout << nums[i];
        if (i < nums.size() - 1) cout << ",";
    }
    cout << "]\n";
    
    switch(approach) {
        case 1: {
            Solution1 sol;
            result = sol.permuteUnique(nums);
            cout << "Approach 1 (Backtracking with Skip): ";
            break;
        }
        case 2: {
            Solution2 sol;
            result = sol.permuteUnique(nums);
            cout << "Approach 2 (Swap with Set): ";
            break;
        }
        case 3: {
            Solution3 sol;
            result = sol.permuteUnique(nums);
            cout << "Approach 3 (Frequency Map): ";
            break;
        }
        case 4: {
            Solution4 sol;
            result = sol.permuteUnique(nums);
            cout << "Approach 4 (Iterative Insertion): ";
            break;
        }
        case 5: {
            Solution5 sol;
            result = sol.permuteUnique(nums);
            cout << "Approach 5 (Next Permutation): ";
            break;
        }
    }
    
    cout << "Count: " << result.size() << "\n";
    printPermutations(result);
    cout << "\n";
}

int main() {
    // Test Case 1: Standard case with duplicates
    cout << "Test Case 1: Standard case\n";
    vector<int> test1 = {1, 1, 2};
    for (int i = 1; i <= 5; i++) {
        test(test1, i);
    }
    
    // Test Case 2: No duplicates
    cout << "Test Case 2: No duplicates\n";
    vector<int> test2 = {1, 2, 3};
    for (int i = 1; i <= 5; i++) {
        test(test2, i);
    }
    
    // Test Case 3: All duplicates
    cout << "Test Case 3: All duplicates\n";
    vector<int> test3 = {1, 1, 1};
    for (int i = 1; i <= 5; i++) {
        test(test3, i);
    }
    
    // Test Case 4: Single element
    cout << "Test Case 4: Single element\n";
    vector<int> test4 = {1};
    for (int i = 1; i <= 5; i++) {
        test(test4, i);
    }
    
    // Test Case 5: Two elements, one duplicate
    cout << "Test Case 5: Two elements\n";
    vector<int> test5 = {1, 1};
    for (int i = 1; i <= 5; i++) {
        test(test5, i);
    }
    
    return 0;
}

/*
 * COMPLEXITY ANALYSIS SUMMARY:
 * 
 * Approach 1 (Backtracking with Skip):
 * - Time: O(n! * n) - n! permutations, O(n) to copy
 * - Space: O(n) - recursion + used array
 * - Best for: Most efficient, clean code
 * 
 * Approach 2 (Swap with Set):
 * - Time: O(n! * n)
 * - Space: O(n^2) - set at each recursion level
 * - Best for: Alternative backtracking approach
 * 
 * Approach 3 (Frequency Map):
 * - Time: O(n! * n)
 * - Space: O(n) - frequency map
 * - Best for: When many duplicates
 * 
 * Approach 4 (Iterative Insertion):
 * - Time: O(n! * n^2) - insertion overhead
 * - Space: O(n! * n) - set storage
 * - Best for: Avoiding recursion (less efficient)
 * 
 * Approach 5 (Next Permutation):
 * - Time: O(n! * n)
 * - Space: O(1) - excluding output
 * - Best for: Using STL, lexicographic order
 * 
 * INTERVIEW TIPS:
 * 1. Start with Approach 1 - most common and efficient
 * 2. Key insight: sort first, then skip duplicates correctly
 * 3. Explain the condition: i > 0 && nums[i] == nums[i-1] && !used[i-1]
 * 4. This ensures duplicates are used in order
 * 5. Mention next_permutation as elegant alternative
 * 
 * COMMON MISTAKES:
 * 1. Not sorting the array first
 * 2. Wrong duplicate-skipping condition
 * 3. Forgetting to mark elements as used/unused
 * 4. Using set unnecessarily (inefficient)
 * 5. Not understanding why the skip condition works
 * 
 * FOLLOW-UP QUESTIONS:
 * 1. How to generate only k-length permutations? (Modify base case)
 * 2. Can you generate permutations in lexicographic order? (Use Approach 5)
 * 3. What if we want to count permutations without generating? (Use formula)
 * 4. How to handle very large arrays? (Consider memory constraints)
 * 5. Can you generate next permutation in-place? (Yes, O(n) algorithm)
 * 
 * RELATED PROBLEMS:
 * - Permutations (without duplicates)
 * - Next Permutation
 * - Subsets II
 * - Combination Sum II
 * - Palindrome Permutation II
 * 
 * KEY INSIGHTS:
 * 1. Sorting groups duplicates together
 * 2. Skip condition ensures duplicates used in order
 * 3. Used array tracks which elements are in current permutation
 * 4. Frequency map is elegant for many duplicates
 * 5. Next permutation provides lexicographic order
 */

// Made with Bob

File: src/data/tutorials/069-permutations-ii.js
/**
 * Tutorial content for #069 — Permutations II
 * Keyed by approach name (must match the name parsed from the solution file).
 */
export const tutorial = {
  problem: {
    statement: `Given a collection of integers that might contain duplicates, return all unique permutations. Two permutations are duplicates if they contain the same elements in the same order.`,
    example: `nums = [1, 1, 2]\n→ [1,1,2], [1,2,1], [2,1,1]\n✅ Answer: 3 unique permutations`,
    keyInsight: `Sort first. During backtracking, if the current element equals the previous one AND the previous one has NOT been used yet (meaning we're at the same depth-level, not in an ongoing branch), skip it to avoid generating a duplicate permutation.`,
  },

  approaches: {
    'Backtracking with Sorting and Skip Duplicates': {
      intuition: `Sort the array. Use a "used" boolean array. For each position, try every unused element. Skip if the element equals the previous element AND the previous element is not currently in use — this ensures each duplicate value is placed in sequence, never in parallel.`,
      steps: [
        `Sort nums.`,
        `Backtrack(current, used):`,
        `  If current.size() == n: add copy; return.`,
        `  For i from 0 to n-1:`,
        `    If used[i]: skip.`,
        `    If i > 0 AND nums[i]==nums[i-1] AND NOT used[i-1]: skip (duplicate branch).`,
        `    Mark used[i]=true, add nums[i], recurse, unmark, remove last.`,
      ],
      example: `nums = [1, 1, 2]  sorted\n\ncurrent=[], used=[F,F,F]\ni=0 (1, not used): mark, [1] ...\ni=1 (1, dup of i=0 AND used[0]=F → SKIP)\ni=2 (2): [2] ...\n\nFrom [1] (used[0]=T):\n  i=1 (1, used[0]=T so NOT skipped): [1,1] → [1,1,2] ✓\n  i=2 (2): [1,2] → [1,2,1] ✓\n\nFrom [2] (used[2]=T):\n  i=0 (1): [2,1] → [2,1,1] ✓\n  i=1 (1, dup of i=0, used[0]=F → SKIP)\n✅ Answer: 3 permutations`,
      keyInsight: `O(n! × n) time, O(n) extra space. The "NOT used[i-1]" condition is the key — it prunes duplicate branches at the same recursion level.`,
    },

    'Backtracking with Swap and Set': {
      intuition: `Use the swap-based backtracking (fix element at each position), but guard duplicates with a HashSet at each recursion level. If nums[j] is already in the set for this position, skip the swap — we've already explored that choice.`,
      steps: [
        `Backtrack(start):`,
        `  If start==n: add copy of nums; return.`,
        `  Create a seen set for this level.`,
        `  For j from start to n-1:`,
        `    If nums[j] is in seen: skip.`,
        `    Add nums[j] to seen.`,
        `    Swap nums[start] and nums[j], recurse(start+1), swap back.`,
      ],
      example: `nums = [1, 1, 2]\n\nstart=0, seen={}:\n  j=0 (1): seen={1}; swap noop; recurse(1)\n    start=1, seen={}:\n      j=1 (1): seen={1}; recurse(2) → [1,1,2] ✓\n      j=2 (2): seen={1,2}; swap → [1,2,1]; recurse(2) → [1,2,1] ✓\n  j=1 (1): already in seen → skip\n  j=2 (2): seen={1,2}; swap → [2,1,1]; recurse(1)\n    ... → [2,1,1] ✓\n✅ Answer: 3 permutations`,
      keyInsight: `O(n! × n) time. The per-level HashSet deduplicates without requiring sorting, at the cost of O(n) per level of extra memory.`,
    },

    'Frequency Map Backtracking': {
      intuition: `Build a frequency map of each distinct value. At each position, try placing each distinct value (not each array index). Decrement the count when you place a value, increment when you backtrack. This naturally generates each unique permutation exactly once.`,
      steps: [
        `Build a sorted map of value → count.`,
        `Backtrack(current, countMap):`,
        `  If current.size() == n: add copy; return.`,
        `  For each distinct value v with count > 0:`,
        `    Decrement count[v], add v to current.`,
        `    Recurse.`,
        `    Increment count[v], remove last from current.`,
      ],
      example: `nums = [1, 1, 2] → map = {1:2, 2:1}\n\ncurrent=[]:\n  Try 1: map={1:1,2:1}, current=[1]\n    Try 1: map={1:0,2:1}, current=[1,1]\n      Try 2: map={1:0,2:0}, current=[1,1,2] → add ✓\n    Try 2: map={1:1,2:0}, current=[1,2]\n      Try 1: current=[1,2,1] → add ✓\n  Try 2: map={1:2,2:0}, current=[2]\n    Try 1 twice: current=[2,1,1] → add ✓\n✅ Answer: 3 permutations`,
      keyInsight: `O(n! × k) time where k is distinct values. Avoids sorting and skip-logic; naturally deduplicates by iterating over values, not indices.`,
    },

    'Iterative Insertion with Set': {
      intuition: `Extend Iterative Insertion from Permutations I with a global HashSet. Start from [[nums[0]]]. For each new number, insert it at all positions in all existing permutations, but only add to results if that permutation hasn't been seen before.`,
      steps: [
        `result = [[nums[0]]], seen = {[nums[0]]}.`,
        `For i from 1 to n-1:`,
        `  For each existing perm:`,
        `    For each position j from 0 to perm.length:`,
        `      Create new perm with nums[i] inserted at j.`,
        `      If not in seen: add to new result, add to seen.`,
        `  result = new result.`,
        `Return result.`,
      ],
      example: `nums = [1, 1, 2]\n\nStart: [[1]]\nInsert 1: [1,1],[1,1] (pos 0 and 1, but [1,1] is duplicate) → {[1,1]}\nInsert 2 into [1,1]: [2,1,1],[1,2,1],[1,1,2] → all new\nResult: [[2,1,1],[1,2,1],[1,1,2]]\n✅ Answer: 3 permutations`,
      keyInsight: `O(n! × n) time. Iterative approach — no recursion — but requires a HashSet for deduplication which adds memory and hashing overhead.`,
    },

    'Next Permutation': {
      intuition: `Sort nums, add it to results, then repeatedly generate the next lexicographic permutation until the sequence wraps back to the sorted order. This produces all unique permutations in sorted order, naturally handling duplicates since identical permutations are never generated as "next" states.`,
      steps: [
        `Sort nums; count total = n! / (freq1! × freq2! × ...).`,
        `Add sorted nums to results.`,
        `Repeat (total - 1) times:`,
        `  Find largest i where nums[i] < nums[i+1].`,
        `  Find largest j where nums[j] > nums[i]; swap.`,
        `  Reverse suffix starting at i+1.`,
        `  Add copy.`,
        `Return results.`,
      ],
      example: `nums = [1, 1, 2]  sorted\n\n[1,1,2] → add\nNext: i=1 (nums[1]=1<nums[2]=2), j=2; swap → [1,2,1]; reverse from 2 → [1,2,1] → add\nNext: i=0 (nums[0]=1<nums[1]=2), j=2; swap → [2,1,1]; reverse from 1 → [2,1,1] → add\nWraps back to [1,1,2] → stop\n✅ Answer: 3 permutations`,
      keyInsight: `O(n! × n) time. Generates permutations in lexicographic order. Duplicates in the input never produce duplicate permutations because "next permutation" skips them structurally.`,
    },
  },
}
